File: examples/nodejs/02_advanced_s3_failover.ts
/**
 * Arkilian Node.js / TypeScript SDK - Advanced S3 Replication & Disaster Recovery
 * 
 * Demonstrates:
 * 1. Cloud-native WAL streaming to S3/MinIO
 * 2. Live telemetry monitoring: health status, health flags, outbox queue depth
 * 3. Explicit WAL flush to ensure zero lag
 * 4. Simulating complete primary host loss (local DB deletion)
 * 5. Instant cold-start disaster recovery via Arkilian.hydrateS3()
 * 6. Validating data integrity on the recovered instance
 */

import * as fs from 'node:fs';
import Arkilian, { S3Config } from 'arkilian';

interface LedgerEntry {
  entry_id: number;
  account_id: string;
  delta: number;
  note: string;
  timestamp?: string;
}

interface CountResult {
  count: number;
}

const S3_CONFIG: S3Config = {
  endpoint: process.env.ARKILIAN_S3_ENDPOINT || 'http://127.0.0.1:9000',
  bucket: process.env.ARKILIAN_S3_BUCKET || 'arkilian-test-bucket',
  region: process.env.ARKILIAN_S3_REGION || 'us-east-1',
  accessKey: process.env.ARKILIAN_S3_ACCESS_KEY || 'minioadmin',
  secretKey: process.env.ARKILIAN_S3_SECRET_KEY || 'minioadmin',
  prefix: process.env.ARKILIAN_S3_PREFIX || 'nodejs-ts-demo',
};

// Configure environment for the Arkilian C engine & background sidecar
process.env.ARKILIAN_ENABLE_BACKUP = '1';
process.env.ARKILIAN_S3_ENDPOINT = S3_CONFIG.endpoint;
process.env.ARKILIAN_S3_BUCKET = S3_CONFIG.bucket;
process.env.ARKILIAN_S3_REGION = S3_CONFIG.region;
process.env.ARKILIAN_S3_ACCESS_KEY = S3_CONFIG.accessKey;
process.env.ARKILIAN_S3_SECRET_KEY = S3_CONFIG.secretKey;
process.env.ARKILIAN_S3_PREFIX = S3_CONFIG.prefix;
process.env.ARKILIAN_MANIFEST_HMAC_KEY = 'super-secret-hmac-key-for-manifests';
process.env.ARKILIAN_CHUNK_INTERVAL_SEC = '1';
process.env.ARKILIAN_MANIFEST_INTERVAL_SEC = '1';

const PRIMARY_DB = './primary_production.sqlite';
const RESTORED_DB = './hydrated_recovery.sqlite';

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function cleanupFile(filePath: string): void {
  for (const ext of ['', '-wal', '-shm']) {
    const p = filePath + ext;
    if (fs.existsSync(p)) fs.unlinkSync(p);
  }
}

async function main(): Promise<void> {
  console.log('=== Arkilian TypeScript SDK: Advanced S3 Streaming & Cold-Start Hydration ===\n');
  console.log(`[Config] Target S3 Endpoint: ${S3_CONFIG.endpoint}`);
  console.log(`[Config] Bucket: ${S3_CONFIG.bucket} | Prefix: ${S3_CONFIG.prefix}\n`);

  cleanupFile(PRIMARY_DB);
  cleanupFile(RESTORED_DB);

  // -------------------------------------------------------------
  // Phase 1: Primary Database Workload with Real-Time Streaming
  // -------------------------------------------------------------
  console.log('[Phase 1] Initializing Primary Node with S3 WAL Streaming...');
  const db = new Arkilian(PRIMARY_DB);

  // Verify sidecar telemetry
  console.log(`  Telemetry -> Sidecar Healthy: ${db.backupHealthy}`);
  console.log(`  Telemetry -> Health Flags: 0x${db.backupHealthFlags.toString(16)}`);
  console.log(`  Telemetry -> Queue Depth: ${db.backupQueueDepth}`);

  console.log('\n[Phase 1] Creating schema and streaming financial ledger entries...');
  db.exec(`
    CREATE TABLE IF NOT EXISTS ledger (
      entry_id INTEGER PRIMARY KEY AUTOINCREMENT,
      account_id TEXT NOT NULL,
      delta REAL NOT NULL,
      note TEXT NOT NULL,
      timestamp DATETIME DEFAULT CURRENT_TIMESTAMP
    );
  `);

  // Insert 20 transactions
  const totalTransactions = 20;
  const insertStmt = 'INSERT INTO ledger (account_id, delta, note) VALUES (?, ?, ?)';

  db.transaction(() => {
    for (let i = 1; i <= totalTransactions; i++) {
      const delta = (i * 12.5).toFixed(2);
      db.run(insertStmt, [`acc_${i % 5}`, parseFloat(delta), `Transaction batch item #${i}`]);
    }
  });
  console.log(`  ✓ Successfully committed ${totalTransactions} ledger entries.`);

  // Flush WAL chunk to outbox & allow sidecar worker to ship to S3
  console.log('\n[Phase 1] Triggering explicit WAL flush to S3 sidecar...');
  db.walFlush();

  // Poll sidecar queue until it drains and uploads chunk to S3
  console.log('  Waiting for S3 sidecar outbox to drain...');
  for (let attempt = 0; attempt < 20; attempt++) {
    await sleep(500);
    if (db.backupQueueDepth === 0) {
      break;
    }
  }
  // Give manifest publish a brief moment to finish commit
  await sleep(1500);

  console.log(`  Telemetry -> Queue Depth: ${db.backupQueueDepth}`);
  console.log(`  Telemetry -> Sidecar Healthy: ${db.backupHealthy}`);

  // Verify records count
  const [primaryCount] = db.all<CountResult>('SELECT COUNT(*) as count FROM ledger');
  console.log(`  Primary DB confirmed record count: ${primaryCount.count}`);

  console.log('[Phase 1] Closing primary database connection.');
  db.close();

  // -------------------------------------------------------------
  // Phase 2: Disaster Simulation (Host Catastrophe)
  // -------------------------------------------------------------
  console.log('\n[Phase 2] SIMULATING DISASTER: Primary host destroyed!');
  console.log(`  Purging local files: ${PRIMARY_DB}...`);
  cleanupFile(PRIMARY_DB);
  console.log('  ✓ Primary local database destroyed. Zero local state remains.');

  // -------------------------------------------------------------
  // Phase 3: Cold-Start Hydration from S3
  // -------------------------------------------------------------
  console.log('\n[Phase 3] Starting Cold-Start Hydration from MinIO S3...');
  console.log(`  Target path: ${RESTORED_DB}`);

  try {
    Arkilian.hydrateS3(RESTORED_DB, S3_CONFIG);
    console.log('  ✓ Hydration completed successfully!');
  } catch (err) {
    console.error('  ✗ Hydration failed:', err);
    process.exit(1);
  }

  // -------------------------------------------------------------
  // Phase 4: Data Parity & Integrity Verification
  // -------------------------------------------------------------
  console.log('\n[Phase 4] Opening Hydrated Database to verify integrity...');
  process.env.ARKILIAN_ENABLE_BACKUP = '0';
  const recoveredDb = new Arkilian(RESTORED_DB);

  const [recoveredCount] = recoveredDb.all<CountResult>('SELECT COUNT(*) as count FROM ledger');
  console.log(`  Recovered DB record count: ${recoveredCount.count}`);

  if (recoveredCount.count !== primaryCount.count) {
    console.error(`  ✗ Data mismatch! Expected ${primaryCount.count}, got ${recoveredCount.count}`);
    process.exit(1);
  }

  console.log('\n  Sample records from recovered instance:');
  const sampleRows = recoveredDb.all<LedgerEntry>(
    'SELECT entry_id, account_id, delta, note FROM ledger LIMIT 5'
  );
  console.table(sampleRows);

  recoveredDb.close();
  cleanupFile(RESTORED_DB);
  console.log('\n=== Disaster Recovery & Verification Finished Successfully! ===');
}

main().catch((err) => {
  console.error('Fatal execution error:', err);
  process.exit(1);
});

File: examples/rust/src/bin/02_advanced_s3_failover.rs
//! Arkilian Rust SDK - Advanced S3 Streaming & Cold-Start Hydration
//!
//! Demonstrates:
//! 1. Continuous WAL streaming to MinIO / AWS S3
//! 2. Telemetry inspection (sidecar health, bitmask health flags, outbox queue depth)
//! 3. Zero-lag WAL flushes via db.wal_flush()
//! 4. Primary host disaster simulation (deleting local database files)
//! 5. Cold-start disaster recovery via arkilian::hydrate_s3()
//! 6. Data parity and integrity validation on the recovered database

use std::env;
use std::fs;
use std::path::Path;
use std::thread;
use std::time::Duration;
use arkilian::{hydrate_s3, Database, S3Config, SQLITE_ROW};

const PRIMARY_DB: &str = "primary_production.sqlite";
const RESTORED_DB: &str = "hydrated_recovery.sqlite";

fn cleanup_db(path: &str) {
    for ext in &["", "-wal", "-shm"] {
        let file = format!("{}{}", path, ext);
        if Path::new(&file).exists() {
            let _ = fs::remove_file(file);
        }
    }
}

fn get_env(key: &str, fallback: &str) -> String {
    env::var(key).unwrap_or_else(|_| fallback.to_string())
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("=== Arkilian Rust SDK: Advanced S3 Streaming & Cold-Start Hydration ===\n");

    let endpoint = get_env("ARKILIAN_S3_ENDPOINT", "http://127.0.0.1:9000");
    let bucket = get_env("ARKILIAN_S3_BUCKET", "arkilian-test-bucket");
    let region = get_env("ARKILIAN_S3_REGION", "us-east-1");
    let access_key = get_env("ARKILIAN_S3_ACCESS_KEY", "minioadmin");
    let secret_key = get_env("ARKILIAN_S3_SECRET_KEY", "minioadmin");
    let prefix = get_env("ARKILIAN_S3_PREFIX", "rust-demo");
    let hmac_key = get_env("ARKILIAN_MANIFEST_HMAC_KEY", "super-secret-hmac-key-for-manifests");

    // Configure environment variables for the Arkilian C sidecar worker
    env::set_var("ARKILIAN_ENABLE_BACKUP", "1");
    env::set_var("ARKILIAN_S3_ENDPOINT", &endpoint);
    env::set_var("ARKILIAN_S3_BUCKET", &bucket);
    env::set_var("ARKILIAN_S3_REGION", &region);
    env::set_var("ARKILIAN_S3_ACCESS_KEY", &access_key);
    env::set_var("ARKILIAN_S3_SECRET_KEY", &secret_key);
    env::set_var("ARKILIAN_S3_PREFIX", &prefix);
    env::set_var("ARKILIAN_MANIFEST_HMAC_KEY", &hmac_key);
    env::set_var("ARKILIAN_CHUNK_INTERVAL_SEC", "1");
    env::set_var("ARKILIAN_MANIFEST_INTERVAL_SEC", "1");

    println!("[Config] Target S3 Endpoint: {}", endpoint);
    println!("[Config] Bucket: {} | Prefix: {}\n", bucket, prefix);

    cleanup_db(PRIMARY_DB);
    cleanup_db(RESTORED_DB);

    // -------------------------------------------------------------
    // Phase 1: Primary Database Workload with Real-Time S3 Streaming
    // -------------------------------------------------------------
    println!("[Phase 1] Opening Primary Node with S3 WAL Streaming...");
    let mut db = Database::new(PRIMARY_DB).map_err(|e| format!("Open error: {}", e))?;

    // Verify sidecar telemetry
    println!("  Telemetry -> Sidecar Healthy: {}", db.backup_is_healthy());
    println!("  Telemetry -> Health Flags: 0x{:x}", db.backup_health_flags());
    println!("  Telemetry -> Queue Depth: {}", db.backup_queue_depth());

    println!("\n[Phase 1] Creating schema and streaming mission event logs...");
    let schema_sql = r#"
        CREATE TABLE IF NOT EXISTS mission_events (
            event_id INTEGER PRIMARY KEY AUTOINCREMENT,
            module_name TEXT NOT NULL,
            severity TEXT NOT NULL,
            payload_value REAL NOT NULL,
            timestamp DATETIME DEFAULT CURRENT_TIMESTAMP
        );
    "#;
    db.exec(schema_sql).map_err(|e| format!("Schema error: {}", e))?;

    let total_records = 30;
    let insert_sql = "INSERT INTO mission_events (module_name, severity, payload_value) VALUES (?, ?, ?);";

    db.begin().map_err(|e| format!("Begin error: {}", e))?;
    for i in 1..=total_records {
        db.prepare(insert_sql).map_err(|e| format!("Prepare error: {}", e))?;
        db.bind_text(1, &format!("telemetry_subsystem_{}", i % 5))?;
        let severity = if i % 7 == 0 { "CRITICAL" } else if i % 3 == 0 { "WARN" } else { "INFO" };
        db.bind_text(2, severity)?;
        db.bind_double(3, (i as f64) * 14.25)?;
        let _ = db.step();
        db.finalize().map_err(|e| format!("Finalize error: {}", e))?;
    }
    db.commit().map_err(|e| format!("Commit error: {}", e))?;
    println!("  ✓ Committed {} mission events in an atomic transaction.", total_records);

    // Explicit WAL flush: wake up flush thread and drain to S3
    println!("\n[Phase 1] Triggering explicit WAL flush to S3 sidecar...");
    db.wal_flush();

    println!("  Waiting for S3 sidecar outbox to drain...");
    for _ in 0..20 {
        thread::sleep(Duration::from_millis(500));
        if db.backup_queue_depth() == 0 {
            break;
        }
    }
    // Allow manifest publish to finalize
    thread::sleep(Duration::from_millis(1500));

    println!("  Telemetry -> Queue Depth: {}", db.backup_queue_depth());
    println!("  Telemetry -> Sidecar Healthy: {}", db.backup_is_healthy());

    // Record count on primary
    db.prepare("SELECT COUNT(*) FROM mission_events;").map_err(|e| format!("Prepare error: {}", e))?;
    let _ = db.step();
    let primary_count = db.column_int64(0);
    db.finalize().map_err(|e| format!("Finalize error: {}", e))?;
    println!("  Primary DB confirmed record count: {}", primary_count);

    println!("[Phase 1] Closing primary database connection.");
    db.close();

    // -------------------------------------------------------------
    // Phase 2: Disaster Simulation
    // -------------------------------------------------------------
    println!("\n[Phase 2] SIMULATING DISASTER: Destroying primary host!");
    println!("  Purging local files: {}...", PRIMARY_DB);
    cleanup_db(PRIMARY_DB);
    println!("  ✓ Primary local database destroyed. Zero local state remains.");

    // -------------------------------------------------------------
    // Phase 3: Cold-Start Hydration from MinIO S3
    // -------------------------------------------------------------
    println!("\n[Phase 3] Starting Cold-Start Hydration from MinIO S3...");
    println!("  Target path: {}", RESTORED_DB);

    let s3_config = S3Config {
        endpoint: endpoint.clone(),
        bucket: bucket.clone(),
        region: region.clone(),
        access_key_id: access_key.clone(),
        secret_access_key: secret_key.clone(),
        session_token: None,
        use_ssl: false,
        timeout_ms: 15_000,
    };

    if let Err(e) = hydrate_s3(RESTORED_DB, &prefix, &s3_config) {
        eprintln!("  ✗ Hydration failed: {}", e);
        std::process::exit(1);
    }
    println!("  ✓ Hydration completed successfully!");

    // -------------------------------------------------------------
    // Phase 4: Data Parity & Integrity Verification
    // -------------------------------------------------------------
    println!("\n[Phase 4] Opening Hydrated Database to verify integrity...");
    env::set_var("ARKILIAN_ENABLE_BACKUP", "0");
    let mut recovered_db = Database::new(RESTORED_DB).map_err(|e| format!("Open error: {}", e))?;

    recovered_db.prepare("SELECT COUNT(*) FROM mission_events;").map_err(|e| format!("Prepare error: {}", e))?;
    let _ = recovered_db.step();
    let recovered_count = recovered_db.column_int64(0);
    recovered_db.finalize().map_err(|e| format!("Finalize error: {}", e))?;
    println!("  Recovered DB record count: {}", recovered_count);

    if recovered_count != primary_count {
        eprintln!("  ✗ Data mismatch! Expected {}, got {}", primary_count, recovered_count);
        std::process::exit(1);
    }

    println!("\n  Sample records from recovered instance:");
    recovered_db.prepare("SELECT event_id, module_name, severity, payload_value FROM mission_events LIMIT 5;")
        .map_err(|e| format!("Prepare error: {}", e))?;
    while recovered_db.step() == SQLITE_ROW {
        println!("    - Event #{:<2} | Module: {:<22} | Severity: {:<8} | Value: {:>8.2}",
            recovered_db.column_int64(0),
            recovered_db.column_text(1).unwrap_or_default(),
            recovered_db.column_text(2).unwrap_or_default(),
            recovered_db.column_double(3));
    }
    recovered_db.finalize().map_err(|e| format!("Finalize error: {}", e))?;

    recovered_db.close();
    cleanup_db(RESTORED_DB);

    println!("\n=== Disaster Recovery & Verification Finished Successfully! ===");
    Ok(())
}
